File: C++/Solution120.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
#include<queue>
using namespace std;

typedef long long ll;
typedef pair<int, int> pii;

class Solution120 {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int n = triangle.size(), m = triangle.back().size();
        vector<vector<int>> dp(n, vector<int>(m, INT_MAX));
        dp.back() = triangle.back();
        for(int i = n - 2; i >= 0; i--){
            for(int j = 0; j <= i; j++){
                for(int dep = 0; dep <= 1; dep++){
                    if(dep >= 0 && dep < m){
                        dp[i][j] = min(dp[i][j], triangle[i][j] + dp[i + 1][j + dep]);
                    }
                }
            }
        }
        return dp.front().front();
    }
};

File: Java/other/Solution120.java
import java.util.List;
import java.util.LinkedList;

class Solution120 {
    public int minimumTotal(List<List<Integer>> triangle) {
    	int size = triangle.size();
    	int[] num = new int[size];
    	for(int i = 0; i < size; i++){
    		num[i] = triangle.get(size - 1).get(i);
    	}
    	for(int i = size - 2; i >= 0; i--){
    		for(int j = 0; j <= i; j++){
    			num[j] = Math.min(num[j], num[j + 1]) + triangle.get(i).get(j);
    		}
    	}
    	return num[0];
    }
}